subsystem/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+
+    fn new(manager: &'a Manager) -> Self {
+        Self {
+            manager,
+            subsystem_name: Self::DEFAULT_SUBSYSTEM_NAME,
+            call_queue_capacity: Self::DEFAULT_CALL_QUEUE_CAPACITY,
+        }
+    }
+
+    /// Set call queue capacity
+    pub fn with_call_queue_capacity(mut self, new_cap: usize) -> Self {
+        self.call_queue_capacity = new_cap;
+        self
+    }
+
+    /// Set subsystem name
+    pub fn with_name(mut self, new_name: &'static str) -> Self {
+        self.subsystem_name = new_name;
+        self
+    }
+
+    /// Start the subsystem.
+    ///
+    /// A subsystem has to handle shutdown and call requests. It can also react to external IO
+    /// events. If the subsystem handles *only* calls and shutdown requests without interaction
+    /// with any additional IO, use the [Manager::start_passive] convenience method.
+    ///
+    /// A typical skeleton of a subsystem looks like this:
+    /// ```no_run
+    /// # let manager = subsystem::Manager::new("app");
+    /// let subsystem = manager.start("my-subsystem", |mut call, mut shutdown| async move {
+    ///     loop {
+    ///         tokio::select! {
+    ///             // Shutdown received, break out of the loop.
+    ///             () = shutdown.recv() => { break; }
+    ///             // Handle calls. An object representing the subsystem is passed in.
+    ///             func = call.recv() => { func(todo!("put an argument here")); }
+    ///             // Handle any other IO events here
+    ///         };
+    ///     }
+    /// });
+    /// # let _ = subsystem.call(|()| ());  // Fix the call type to avoid amnbiguity.
+    /// ```
+    pub fn start<T: 'static + Send, F: 'static + Send + Future<Output = ()>>(
+        self,
+        subsystem: impl 'static + Send + FnOnce(CallRequest<T>, ShutdownRequest) -> F,
+    ) -> Subsystem<T> {
+        // Name strings
+        let manager_name = self.manager.name;
+        let subsys_name = self.subsystem_name;
+        // Shutdown-related channels
+        let shutting_down_tx = self.manager.shutting_down_tx.clone();
+        let shutdown_rq = ShutdownRequest(self.manager.shutdown_request_tx.subscribe());
+        // Call related channels
+        let (action_tx, action_rx) = mpsc::channel(self.call_queue_capacity);
+        let call_rq = CallRequest(action_rx);
+
+        task::spawn(async move {
+            log::info!("Subsystem {}/{} started", manager_name, subsys_name);
+
+            // Perform the subsystem task.
+            subsystem(call_rq, shutdown_rq).await;
+
+            // Signal the intent to shut down to the other parts of the application.
+            shutting_down_tx.send(()).await.expect("Subsystem outlived the manager!?");
+
+            log::info!("Subsystem {}/{} terminated", manager_name, subsys_name);
+
+            // Close the channel to signal the completion of the shutdown.
+            std::mem::drop(shutting_down_tx);
```

Comment:
Like I mentioned in our call, I personally prefer not have calls to `std::mem::drop` unless they are strictly necessary, since whenever I see such a call I wonder why it is there. I would remove the call and modify the comment to say something like "shutting_down_tx is dropped at the end of this scope, thus signalling the completion of the shutdown". I understand your desire to be explicit though, just stating my point of view.